Sources/FileSystem/Beta Group/BetaGroupProcessor.swift

```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Itty Bitty Apps Pty Ltd
+
+import Files
+import Foundation
+import Model
+import Yams
+
+public struct BetaGroupProcessor: ResourceProcessor {
+
+    var path: ResourcePath
+
+    public init(path: ResourcePath) {
+        self.path = path
+    }
+
+    public func read() throws -> [BetaGroup] {
+        try getFolder().files.compactMap { (file: File) -> BetaGroup? in
+            if file.extension == "yml" {
+                return Readers
+                    .FileReader<BetaGroup>(format: .yaml)
+                    .read(filePath: file.path)
+            }
+
+            return nil
+        }
+    }
+
+    public func write(groupsWithTesters: [(betaGroup: BetaGroup, testers: [BetaTester])]) throws {
+        deleteFile()
+
+        let betagroups = try groupsWithTesters
+            .map { try write(betaTesters: $0.testers, into: $0.betaGroup) }
+
+        try write(betagroups)
+    }
+
+    @discardableResult
+    func write(_ betaGroups: [BetaGroup]) throws -> [File] {
+        try betaGroups.map { try write($0) }
+    }
+
+    @discardableResult
+    func write(_ betaGroup: BetaGroup) throws -> File {
+        try writeFile(betaGroup)
+    }
+
+    private func write(betaTesters: [BetaTester], into betaGroup: BetaGroup) throws -> BetaGroup {
+
+        var group = betaGroup
+        group.testers = try BetaTesterProcessor(folder: try getFolder())
+            .write(group: group, testers: betaTesters)
+
+        return group
+    }
+
+}
+
+extension BetaGroup: FileProvider {
+    var fileName: String {
+        "\(app.bundleId ?? "")_\(groupName).yml"
```

Comment:
I think we should be normalising the filename a bit better here. `groupName` could be an unsafe filename.  We should turn all non alpha numeric characters into underscores.

```
extension String {
  func filenameSafe() -> String {
    let unsafeFilenameCharacters = CharacterSet(charactersIn: " *?:/\\.")
    return str.components(separatedBy: unsafeFilenameCharacters).joined(separator: "_")
  }
}
```

Reply:
For an account that has a lot of groups it would be better to nest them in subdirectories.  The `app.bundleId` could be a subdirectory rather than a prefix.

Also, surely the `app.bundleId` will be non-nil?  When will it ever be nil?

Reply:
Actually thinking about he normalised/safe filename some more I think when group names are translated to file/path names they should be really cleaned up:

Eg: A group name like `d_App Team (Learn & Adopt)` should be translated to `d-app-team-learn-adopt`. The problem with that however is that removing characters might mean there could be clashes. We might have to come up with some resolution strategy in that case.

Reply:
What I'm trying to suggest here is that the filenames should be names that a developer might choose if they were authoring them themselves.